### DISCUSSION TITLE
A model for task nodes

### DISCUSSION THREAD

**Discussion by mmontone** [neomacs-project/neomacs#89] [Q&A]
Hello.
I would like some input for an experiment I'm trying. I would like to add a node of type "task", so I have this:

```lisp
(in-package :neomacs)

(defmodel task (element)
    ((done :initform nil
           :accessor task-done-p)
     (description :initarg :description
                  :accessor task-description
                  :initform ""
                  :type string)
     (deadline :initarg :deadline
               :accessor task-deadline
               :initform nil)
     (started :initarg :started
              :initform (get-universal-time)
              :accessor task-started))
  (:default-initargs
   :tag-name "div"))

(defun make-task (description &key done)
  (let ((node (make-instance 'task :description description)))
    (append-child node (make-element "input" :type "checkbox" :checked (when done "true")))
    ;;(append-child node (make-element "input" :type "text" :placeholder "Task name..."))
    (append-child node (make-instance 'text-node :text (task-description node)))
    node))

(define-command insert-task (&optional (marker (focus)))
  (insert-nodes marker (make-task "enter task ...")))
```
And my questions:
- Is my implementation approach correct? What would you change?
- Is it possible to have undeletable nodes? I would like the user to be able to delete the whole task node, but not parts of it (like delete only the checkbox or description).
- How can I bind an action to the checkbox? (and update the task status from that)

**Comment by kchanqvq** [comment]
> * Is my implementation approach correct? What would you change?

The use of CLOS element subclass and slots is interesting. What I have written so far has been (loosely) following the convention to use the DOM (which consist of only `text-node` and `element`) as the single source of truth, so I would put all information into `attribute` of the elements. Using CLOS slots do es seem nicer and it is something I haven't really thought through, so no guarantee such approach will work well at the moment (but it'll be interesting to see how to get it work well!)



> * Is it possible to have undeletable nodes? I would like the user to be able to delete the whole task node, but not parts of it (like delete only the checkbox or description).

`(setf (element 'read-only) t)` makes its immediate descendent level readonly. See generic function `check-read-only`.


> * How can I bind an action to the checkbox? (and update the task status from that)

This requires writing some JavaScript (probably by setting the "onchange" attribute with a JavaScript string). JavaScript can call `electronAPI.send('neomacs',{type:"<type>",…})`, which will generates an event that can be handled by Lisp by setting `(setf (gethash "<type>" *ipc-handler-table*) (lambda (buffer event) …))`. See the `;;; Mouse support` section in `buffer.lisp` for an example.

Obviously the current API is quite rough and rudimentary and we will want to wrap it in a nicer high-level API. At least a `define-ipc-handler` macro. Feel free to share if you have more ideas!

**Reply by mmontone** [reply]
Ok. Thanks for the reply! I will experiment a bit over the week.

> > * Is it possible to have undeletable nodes? I would like the user to be able to delete the whole task node, but not parts of it (like delete only the checkbox or description).
> 
> `(setf (element 'read-only) t)` makes its immediate descendent level readonly. See generic function `check-read-only`.

Note that there's a distinction between something being undeletable and something being read-only. I'd need node to be undeletable, but not read-only. I'm think of the checkbox and descrpition of the task. The user should be able to edit them, but not delete them individually, only the whole task object.

**Reply by kchanqvq** [reply]
I think you can have a task element (maybe `li` or `p`) with two children: a checkbox element and a description element (maybe a `span`), then make the task read-only but both children editable. Does that achieve the desired behavior? Note that the read-only attribute only affect an element's immediate inner level (i.e. no immediate child elements or characters can be added or deleted, but child elements themselves can be edited).

**Reply by mmontone** [reply]
It probably does. Yes. I'll try. Thanks.

**Comment by kchanqvq** [comment]
I thought about this for a while and I think I have a better way of handling this, rather than asking you to bind JavaScript event handlers. I can let the `;;; Mouse support` facility do its thing (i.e., moving focus to clicked position), then emit a `mouse-down` **key** event. Then you can add a `'keymap` attribute to the button element, which can bind the `mouse-down` event (Neomacs has node-local keymaps, with inner ones taken precedence over outer ones, then over buffer keymaps, sort of reinventing JavaScript's event "bubble up" phase). This way everything can be done in Lisp world and IMO much cleaner.

One question remain: When should the `;;; Mouse support` facility capture JavaScript event (it registers an event handler on the entire `document` element)? There are two choice:
- Bubble down phase. In this way `;;; Mouse support` handler will capture all mouse events by default and registering `mouse-down` in node-local keymaps works out of the box. However this is somewhat less flexible, because it completely prevents JavaScript event handlers from running.
- Bubble up phase (the current implementation). In this way JavaScript event handlers registered on elements take precedence over `;;; Mouse support` handler. This is more flexible, because JavaScript event handlers get a chance to run if you need it, but that also means you have to explicitly register a handler to override the default (e.g. `onclick="event.preventDefault()"` for a button) to let events reach `;;; Mouse support` handler and subsequently Lisp keymaps.

Any ideas?

**Reply by mmontone** [reply]
> Neomacs has node-local keymaps, with inner ones taken precedence over outer ones, then over buffer keymaps, sort of reinventing JavaScript's event "bubble up" phase

This is interesting to know. I'll try to use it to toggle checkboxes using keyboard.

About the event phase, I don't understand the intricacies well enough to be helpful.

**Comment by kchanqvq** [comment]
Besides, are you experimenting with a TODO/task system? For the general problem I'd consider allowing putting checkbox anywhere, which is superset of both Org behavior (checkbox in headings) and Markdown behavior (checkbox in items). Then something like `org-agenda` would gather all elements (or maybe innermost block-level element) enclosing the checkbox. Just my 5 cents :)

Also be cautious if you are dealing with `li`. Make sure you only ever try to put them inside an enclosing `ul` or `ol` at any moment. I document this in https://neomacs-project.github.io/doc/dom.html#notes-on-dom-consistency. The gist is `li` without enclosing `ul` or `ol` is invalid HTML and the browser engine will silently fix it by enclosing it, which Neomacs won't know about and will cause DOM inconsistency (corrupted document).

**Comment by mmontone** [comment]
I'm running some little experiments for now, with things that I'm interested in for future modes (the tasks, the buttons, the event handling, etc), to test current Neomacs capabilities and get acquainted with Neomacs implementation. So not permanent things just yet :)
